<!DOCTYPE html>
<html lang="en-us">
<head>

  <meta charset="utf-8">
  <meta name="viewport" content="width=device-width, initial-scale=1">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="theme" content="hugo-academic">
  <meta name="generator" content="Hugo 0.34" />
  <meta name="author" content="Spencer Woody">

  
  
  
  
    
      
    
  
  <meta name="description" content="Introduction The model Implementation in R Read in the data Plot the data  Appendix: R session information    This project is currently a WIP (work in progress)
Introduction Hierarchical Gaussian process regression with application to drosophila microarray data, implemented in R and visualized using ggplot2.
Based on a paper by Hensman, Lawrence, and Rattray.
Let \(i\) be the subscript for clusters of genes, \(n\) in the subscript genes, \(r\) is the subscript for replicates.">

  
  <link rel="alternate" hreflang="en-us" href="/post/hgp/">

  


  

  
  
  <meta name="theme-color" content="#0095eb">
  
  
  
  
    
  
  
    
    
    <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
    
  
  
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/css/bootstrap.min.css" integrity="sha512-6MXa8B6uaO18Hid6blRMetEIoPqHf7Ux1tnyIQdpt9qI5OACx7C+O3IVTr98vwGnlcg0LOLa02i9Y1HpVhlfiw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/academicons/1.8.1/css/academicons.min.css" integrity="sha512-NThgw3XKQ1absAahW6to7Ey42uycrVvfNfyjqcFNgCmOCQ5AR4AO0SiXrN+8ZtYeappp56lk1WtvjVmEa+VR6A==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha512-SfTiTlX6kk+qitfevl/7LibUOeJWlt9rbyDn92a1DqWOw9vWG2MFoays0sgObmWazO5BQPiFucnnEAjpAB+/Sw==" crossorigin="anonymous">
  <link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.css" integrity="sha256-ygkqlh3CYSUri3LhQxzdcm0n1EQvH2Y+U5S2idbLtxs=" crossorigin="anonymous">
  
  
  
  
  <link rel="stylesheet" href="//fonts.googleapis.com/css?family=Montserrat:400,700%7cRoboto:400,400italic,700%7cRoboto&#43;Mono">
  
  <link rel="stylesheet" href="/styles.css">
  

  

  
  <link rel="alternate" href="/index.xml" type="application/rss+xml" title="Spencer Woody">
  <link rel="feed" href="/index.xml" type="application/rss+xml" title="Spencer Woody">
  

  <link rel="manifest" href="/site.webmanifest">
  <link rel="icon" type="image/png" href="/img/icon.png">
  <link rel="apple-touch-icon" type="image/png" href="/img/icon-192.png">

  <link rel="canonical" href="/post/hgp/">

  <meta property="twitter:card" content="summary_large_image">
  
  <meta property="twitter:site" content="@spencerwoody">
  <meta property="twitter:creator" content="@spencerwoody">
  
  <meta property="og:site_name" content="Spencer Woody">
  <meta property="og:url" content="/post/hgp/">
  <meta property="og:title" content="Hierarchical nonparametric regression | Spencer Woody">
  <meta property="og:description" content="Introduction The model Implementation in R Read in the data Plot the data  Appendix: R session information    This project is currently a WIP (work in progress)
Introduction Hierarchical Gaussian process regression with application to drosophila microarray data, implemented in R and visualized using ggplot2.
Based on a paper by Hensman, Lawrence, and Rattray.
Let \(i\) be the subscript for clusters of genes, \(n\) in the subscript genes, \(r\) is the subscript for replicates.">
  <meta property="og:locale" content="en-us">
  
  <meta property="article:published_time" content="2018-01-30T00:00:00&#43;00:00">
  
  <meta property="article:modified_time" content="2018-01-30T00:00:00&#43;00:00">
  

  

  <title>Hierarchical nonparametric regression | Spencer Woody</title>

</head>
<body id="top" data-spy="scroll" data-target="#toc" data-offset="71" >

<nav class="navbar navbar-default navbar-fixed-top" id="navbar-main">
  <div class="container">

    
    <div class="navbar-header">
      
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse"
              data-target=".navbar-collapse" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      
      <a class="navbar-brand" href="/">Spencer Woody</a>
    </div>

    
    <div class="collapse navbar-collapse">

      
      
      <ul class="nav navbar-nav navbar-right">
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#about">
            
            <span>Home</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/post">
            
            <span>Blog</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#projects">
            
            <span>Projects</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/post/-my-favorite-things/">
            
            <span>Resources</span>
            
          </a>
        </li>

        
        

        
        
        
        
        
          
        

        <li class="nav-item">
          <a href="/#contact">
            
            <span>Contact</span>
            
          </a>
        </li>

        
        
      

      
      </ul>

    </div>
  </div>
</nav>


<article class="article" itemscope itemtype="http://schema.org/Article">

  


  <div class="article-container">
    <h1 itemprop="name">Hierarchical nonparametric regression</h1>

    

<div class="article-metadata">

  <span class="article-date">
    
    <time datetime="2018-01-30 00:00:00 &#43;0000 UTC" itemprop="datePublished dateModified">
      Jan 30, 2018
    </time>
  </span>
  <span itemscope itemprop="author publisher" itemtype="http://schema.org/Person">
    <meta itemprop="name" content="Spencer Woody">
  </span>

  

  
  

  
  
  
  <span class="middot-divider"></span>
  <span class="article-categories">
    <i class="fa fa-folder"></i>
    
    <a href="/categories/project">Project</a
    >, 
    
    <a href="/categories/r">R</a
    >, 
    
    <a href="/categories/wip">wip</a
    >
    
  </span>
  
  

  
  
<div class="share-box" aria-hidden="true">
  <ul class="share">
    <li>
      <a class="twitter"
         href="https://twitter.com/intent/tweet?text=Hierarchical%20nonparametric%20regression&amp;url=%2fpost%2fhgp%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-twitter"></i>
      </a>
    </li>
    <li>
      <a class="facebook"
         href="https://www.facebook.com/sharer.php?u=%2fpost%2fhgp%2f"
         target="_blank" rel="noopener">
        <i class="fa fa-facebook"></i>
      </a>
    </li>
    <li>
      <a class="linkedin"
         href="https://www.linkedin.com/shareArticle?mini=true&amp;url=%2fpost%2fhgp%2f&amp;title=Hierarchical%20nonparametric%20regression"
         target="_blank" rel="noopener">
        <i class="fa fa-linkedin"></i>
      </a>
    </li>
    <li>
      <a class="weibo"
         href="http://service.weibo.com/share/share.php?url=%2fpost%2fhgp%2f&amp;title=Hierarchical%20nonparametric%20regression"
         target="_blank" rel="noopener">
        <i class="fa fa-weibo"></i>
      </a>
    </li>
    <li>
      <a class="email"
         href="mailto:?subject=Hierarchical%20nonparametric%20regression&amp;body=%2fpost%2fhgp%2f">
        <i class="fa fa-envelope"></i>
      </a>
    </li>
  </ul>
</div>


  

</div>


    <div class="article-style" itemprop="articleBody">
      <div id="TOC">
<ul>
<li><a href="#introduction">Introduction</a><ul>
<li><a href="#the-model">The model</a></li>
<li><a href="#implementation-in-r">Implementation in R</a><ul>
<li><a href="#read-in-the-data">Read in the data</a></li>
<li><a href="#plot-the-data">Plot the data</a></li>
</ul></li>
<li><a href="#appendix-r-session-information">Appendix: R session information</a></li>
</ul></li>
</ul>
</div>

<p><strong>This project is currently a WIP (work in progress)</strong></p>
<div id="introduction" class="section level1">
<h1>Introduction</h1>
<p>Hierarchical Gaussian process regression with application to drosophila microarray data, implemented in R and visualized using <code>ggplot2</code>.</p>
<p>Based on a paper by <a href="https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-252">Hensman, Lawrence, and Rattray</a>.</p>
<p>Let <span class="math inline">\(i\)</span> be the subscript for clusters of genes, <span class="math inline">\(n\)</span> in the subscript genes, <span class="math inline">\(r\)</span> is the subscript for replicates. If gene <span class="math inline">\(n\)</span> belongs to cluster <span class="math inline">\(i\)</span> we denote this with <span class="math inline">\(n \in c_i,\)</span> and <span class="math inline">\(N_i = \# \left\{ n \in c_i \right\}.\)</span> Each gene <span class="math inline">\(n\)</span> has <span class="math inline">\(N_n\)</span> replicates, and each replicate <span class="math inline">\(r\)</span> of gene <span class="math inline">\(n\)</span> has <span class="math inline">\(N_{nr}\)</span> measuremeants across time. Note that because every array of genes is measured all at once, each gene has the same <span class="math inline">\(D = \sum_{r=1}^{N_n} N_{nr}\)</span> total measurements across time for all replicates.</p>
<p>We model gene expression levels as a hierarchical Gaussian process,</p>
<p><span class="math display">\[  \begin{aligned} h_i(t) &amp;\sim \text{GP}(\mathbf{0}, k_h(t, t&#39;)) \\
    g_n(t) &amp;\sim \text{GP}(h_i(t), k_g(t, t&#39;)) \text{ for } n\in c_i \\
    f_{nr}(t) &amp;\sim \text{GP}(g_n(t), k_f(t, t&#39;)), \end{aligned}  \]</span></p>
<p>for some covariance functions <span class="math inline">\(k_h(t, t&#39;)\)</span>, <span class="math inline">\(k_g(t, t&#39;)\)</span>, and <span class="math inline">\(k_f(t, t&#39;)\)</span>.</p>
<p>Our data consist of a small subset of from a time-course DNA microarray experiment with <em>Drosophila melanogaster</em>. The gene expression levels of 14 genes are measured over time.</p>
<div id="the-model" class="section level2">
<h2>The model</h2>
<p>Suppose we have <span class="math inline">\(\mathbf{h}_i\)</span>, <span class="math inline">\(\mathbf{g}_n\)</span>, and <span class="math inline">\(\mathbf{f}_{nr}\)</span>, which are vectors of draws from draws from <span class="math inline">\(h_i(t)\)</span>, <span class="math inline">\(g_n(t)\)</span>, and <span class="math inline">\(f_{nr}(t)\)</span>, respectively, at times <span class="math inline">\(\mathbf{t}_{nr}\)</span>. Define <span class="math inline">\(\mathbf{K}_{f}(\mathbf{t}_{nr}, \mathbf{t}_{nr&#39;})\)</span> to be the <span class="math inline">\(N_{nr} \times N_{nr&#39;}\)</span> matrix such that its <span class="math inline">\((i, j)\)</span> element is <span class="math inline">\(k_f(\mathbf{t}_{nr}[i], \mathbf{t}_{nr&#39;}[j]),\)</span> and similarly define the matrices <span class="math inline">\(\mathbf{K}_{g}(\mathbf{t}_{nr}, \mathbf{t}_{nr&#39;})\)</span> and <span class="math inline">\(\mathbf{K}_{h}(\mathbf{t}_{nr}, \mathbf{t}_{nr&#39;})\)</span>. We model the response vector as</p>
<p><span class="math display">\[ \mathbf{y}_{nr} = \mathbf{f}_{nr} + \mathbf{e},\;\; \mathbf{e} \sim \mathcal{N}(\mathbf{0}, \sigma^2 \mathcal{I}). \]</span></p>
<p>We can see the following conditional distributions,</p>
<p><span class="math display">\[  
\begin{aligned} 
(\mathbf{y}_{nr} | \mathbf{f}_{nr}, \sigma^2) &amp;\sim \mathcal{N}(\mathbf{f}_{nr}, \sigma^2 \mathcal{I}) \\
(\mathbf{f}_{nr} | \mathbf{g}_{n}, \theta_f ) &amp;\sim \mathcal{N}(\mathbf{g}_{n}, \mathbf{K}_{f}(\mathbf{t}_{nr}, \mathbf{t}_{nr})) \\ 
(\mathbf{g}_{n} | \mathbf{h}_{i}, \theta_g ) &amp;\sim \mathcal{N}(\mathbf{h}_{i}, \mathbf{K}_{g}(\mathbf{t}_{nr}, \mathbf{t}_{nr})) \\
(\mathbf{h}_i | \mathbf{t}_{nr}, \theta_h) &amp;\sim \mathcal{N}(\mathbf{0}, \mathbf{K}_h(\mathbf{t}_{nr}, \mathbf{t}_{nr}))
\end{aligned} 
\]</span></p>
<p>We can see that there are two levels of hierarchy here:</p>
<ul>
<li>Replicates within genes</li>
<li>Genes within clusters</li>
</ul>
<p>It is easy to find the marginal likelihood of <span class="math inline">\(\bf{y}_{nr}\)</span> through the properties of the multivariate normal distribution,</p>
<p><span class="math display">\[
(\mathbf{y}_{nr} | \mathbf{t}_{nr}, \theta_f, \theta_g, \theta_f) \sim \mathcal{N}(\mathbf{0}, \mathbf{K}_{f}(\mathbf{t}_{nr}, \mathbf{t}_{nr}) + \mathbf{K}_{g}(\mathbf{t}_{nr}, \mathbf{t}_{nr}) + \mathbf{K}_{h}(\mathbf{t}_{nr}, \mathbf{t}_{nr}) + \sigma^2 \mathcal{I}).
\]</span></p>
</div>
<div id="implementation-in-r" class="section level2">
<h2>Implementation in R</h2>
<p>First we will library in the required packages and read in the data in the format we want, which reflects the hierarchical nature of the data.</p>
<div id="read-in-the-data" class="section level3">
<h3>Read in the data</h3>
<pre class="r"><code>library(dplyr)
library(forcats)
library(ggplot2)
theme_set(theme_bw(base_size = 12))

drosophila &lt;- read.csv(&quot;data/drosophila.csv&quot;)

drosophila &lt;- drosophila %&gt;%
  mutate(gene = fct_inorder(gene))</code></pre>
</div>
<div id="plot-the-data" class="section level3">
<h3>Plot the data</h3>
<pre class="r"><code>allgenes &lt;- drosophila %&gt;% 
    ggplot(aes(time, log2exp)) +
    geom_point(aes(col = group, pch = replicate)) +
    facet_wrap(~gene) +
    labs(title = &quot;Expression Profiles for all Genes&quot;,
         x = &quot;Time&quot;,
         y = &quot;log2-Normalized gene expression&quot;) + 
    scale_colour_brewer(palette = &quot;Dark2&quot;) +
    scale_x_continuous(breaks = seq(0, 12, by = 2)) </code></pre>
<p>See Figure <a href="#fig:facet1">1</a> for example:</p>
<div class="figure"><span id="fig:facet1"></span>
<img src="/post/hgp_files/figure-html/facet1-1.png" alt="faceted plot of all genes" width="672" />
<p class="caption">
Figure 1: faceted plot of all genes
</p>
</div>
</div>
</div>
<div id="appendix-r-session-information" class="section level2">
<h2>Appendix: R session information</h2>
<pre class="r"><code>sessionInfo()
## R version 3.4.2 (2017-09-28)
## Platform: x86_64-apple-darwin15.6.0 (64-bit)
## Running under: macOS High Sierra 10.13.3
## 
## Matrix products: default
## BLAS: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRblas.0.dylib
## LAPACK: /Library/Frameworks/R.framework/Versions/3.4/Resources/lib/libRlapack.dylib
## 
## locale:
## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8
## 
## attached base packages:
## [1] methods   stats     graphics  grDevices utils     datasets  base     
## 
## other attached packages:
## [1] bindrcpp_0.2     ggplot2_2.2.1    forcats_0.2.0    dplyr_0.7.4.9000
## 
## loaded via a namespace (and not attached):
##  [1] Rcpp_0.12.15       knitr_1.18         bindr_0.1         
##  [4] magrittr_1.5       munsell_0.4.3      tidyselect_0.2.3  
##  [7] colorspace_1.3-2   R6_2.2.2           rlang_0.1.6.9003  
## [10] highr_0.6          plyr_1.8.4         stringr_1.2.0     
## [13] tools_3.4.2        grid_3.4.2         gtable_0.2.0      
## [16] xfun_0.1           htmltools_0.3.6    lazyeval_0.2.1    
## [19] yaml_2.1.16        rprojroot_1.3-2    digest_0.6.13     
## [22] assertthat_0.2.0   tibble_1.4.2       bookdown_0.6      
## [25] RColorBrewer_1.1-2 purrr_0.2.4.9000   glue_1.2.0        
## [28] evaluate_0.10.1    rmarkdown_1.8      blogdown_0.5      
## [31] labeling_0.3       stringi_1.1.5      compiler_3.4.2    
## [34] pillar_1.1.0       scales_0.5.0.9000  backports_1.1.2   
## [37] pkgconfig_2.0.1</code></pre>
</div>
</div>

    </div>

    





    
    

    

    
<section id="comments">
  <div id="disqus_thread"></div>
<script>
    var disqus_config = function () {
    
    
    
    };
    (function() {
        if (["localhost", "127.0.0.1"].indexOf(window.location.hostname) != -1) {
            document.getElementById('disqus_thread').innerHTML = 'Disqus comments not available by default when the website is previewed locally.';
            return;
        }
        var d = document, s = d.createElement('script'); s.async = true;
        s.src = '//' + "https-spencerwoody-github-io" + '.disqus.com/embed.js';
        s.setAttribute('data-timestamp', +new Date());
        (d.head || d.body).appendChild(s);
    })();
</script>
<noscript>Please enable JavaScript to view the <a href="https://disqus.com/?ref_noscript">comments powered by Disqus.</a></noscript>
<a href="https://disqus.com" class="dsq-brlink">comments powered by <span class="logo-disqus">Disqus</span></a>
</section>



  </div>
</article>

<footer class="site-footer">
  <div class="container">
    <p class="powered-by">

      &copy; 2018 &middot; 

      Powered by the
      <a href="https://sourcethemes.com/academic/" target="_blank" rel="noopener">Academic theme</a> for
      <a href="https://gohugo.io" target="_blank" rel="noopener">Hugo</a>.

      <span class="pull-right" aria-hidden="true">
        <a href="#" id="back_to_top">
          <span class="button_icon">
            <i class="fa fa-chevron-up fa-2x"></i>
          </span>
        </a>
      </span>

    </p>
  </div>
</footer>


<div id="modal" class="modal fade" role="dialog">
  <div class="modal-dialog">
    <div class="modal-content">
      <div class="modal-header">
        <button type="button" class="close btn-large" data-dismiss="modal">&times;</button>
        <h4 class="modal-title">Cite</h4>
      </div>
      <div>
        <pre><code class="modal-body tex"></code></pre>
      </div>
      <div class="modal-footer">
        <a class="btn btn-primary btn-outline js-copy-cite" href="#" target="_blank">
          <i class="fa fa-copy"></i> Copy
        </a>
        <a class="btn btn-primary btn-outline js-download-cite" href="#" target="_blank">
          <i class="fa fa-download"></i> Download
        </a>
        <div id="modal-error"></div>
      </div>
    </div>
  </div>
</div>

    

    
    

    
    <script async defer src="//maps.googleapis.com/maps/api/js?key=AIzaSyD4b6oh0SPlHfgNrTgw0WliQH-4knxqQPk"></script>
    

    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.2.1/jquery.min.js" integrity="sha512-3P8rXCuGJdNZOnUx/03c1jOTnMn3rP63nBip5gOP2qmUh5YAdVAvFZ1E+QLZZbC1rtMrQb+mah3AfYW11RUrWA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.imagesloaded/4.1.3/imagesloaded.pkgd.min.js" integrity="sha512-umsR78NN0D23AzgoZ11K7raBD+R6hqKojyBZs1w8WvYlsI+QuKRGBx3LFCwhatzBunCjDuJpDHwxD13sLMbpRA==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha512-iztkobsvnjKfAtTNdHkGVjAYTrrtlC7mGp/54c40wowO7LhURYl3gVzzcEqGl/qKXQltJ2HwMrdLcNUdo+N/RQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/jquery.isotope/3.0.4/isotope.pkgd.min.js" integrity="sha512-VDBOIlDbuC4VWxGJNmuFRQ0Li0SKkDpmGyuhAG5LTDLd/dJ/S0WMVxriR2Y+CyPL5gzjpN4f/6iqWVBJlht0tQ==" crossorigin="anonymous"></script>
    <script src="https://cdnjs.cloudflare.com/ajax/libs/fancybox/3.2.5/jquery.fancybox.min.js" integrity="sha256-X5PoE3KU5l+JcX+w09p/wHl9AzK333C4hJ2I9S5mD4M=" crossorigin="anonymous"></script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/gmaps.js/0.4.25/gmaps.min.js" integrity="sha256-7vjlAeb8OaTrCXZkCNun9djzuB2owUsaO72kXaFDBJs=" crossorigin="anonymous"></script>
    
    
    <script src="/js/hugo-academic.js"></script>
    

    
    
      
      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js" integrity="sha256-/BfiIkHlHoVihZdc6TFuj7MmJ0TWcWsMXkeDFwhi0zw=" crossorigin="anonymous"></script>
      

      
      <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/r.min.js"></script>
      

      

      <script>hljs.initHighlightingOnLoad();</script>
    

    
    
    <script type="text/x-mathjax-config">
        MathJax.Hub.Config({ tex2jax: { inlineMath: [['$','$'], ['\\(','\\)']] } });
    </script>
    
    <script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS_CHTML" integrity="sha512-tOav5w1OjvsSJzePRtt2uQPFwBoHt1VZcUq8l8nm5284LEKE9FSJBQryzMBzHxY5P0zRdNqEcpLIRVYFNgu1jw==" crossorigin="anonymous"></script>
    
    

  </body>
</html>

