---
title: "Hierarchical nonparametric regression"
author: "Spencer Woody"
date: 2018-01-30
categories: ["Project", "R", "wip"]
tags: []
summary: Hierarchical Gaussian process regression with application to drosophila microarray data
---

**This project is currently a WIP (work in progress)**

```{r setup, include=FALSE}
knitr::opts_chunk$set(collapse = TRUE)
```

# Introduction

Hierarchical Gaussian process regression with application to drosophila microarray data, implemented in R and visualized using `ggplot2`. 

Based on a paper by [Hensman, Lawrence, and Rattray](https://bmcbioinformatics.biomedcentral.com/articles/10.1186/1471-2105-14-252).

Let $i$ be the subscript for clusters of genes, $n$ in the subscript genes, $r$ is the subscript for replicates. If gene $n$ belongs to cluster $i$ we denote this with $n \in c_i,$ and $N_i = \# \left\{ n \in c_i \right\}.$ Each gene $n$ has $N_n$ replicates, and each replicate $r$ of gene $n$ has $N_{nr}$ measuremeants across time. Note that because every array of genes is measured all at once, each gene has the same $D = \sum_{r=1}^{N_n} N_{nr}$ total measurements across time for all replicates.

We model gene expression levels as a hierarchical Gaussian process,

$$ 	\begin{aligned} h_i(t) &\sim \text{GP}(\mathbf{0}, k_h(t, t')) \\
	g_n(t) &\sim \text{GP}(h_i(t), k_g(t, t')) \text{ for } n\in c_i \\
	f_{nr}(t) &\sim \text{GP}(g_n(t), k_f(t, t')), \end{aligned}  $$
	
	for some covariance functions $k_h(t, t')$, $k_g(t, t')$, and $k_f(t, t')$. 

Our data consist of a small subset of from a time-course DNA microarray experiment with *Drosophila melanogaster*. The gene expression levels of 14 genes are measured over time. 

## The model

Suppose we have $\mathbf{h}_i$, $\mathbf{g}_n$, and $\mathbf{f}_{nr}$, which are vectors of draws from draws from $h_i(t)$, $g_n(t)$, and $f_{nr}(t)$, respectively, at times $\mathbf{t}_{nr}$. Define $\mathbf{K}_{f}(\mathbf{t}_{nr}, \mathbf{t}_{nr'})$ to be the $N_{nr} \times N_{nr'}$ matrix such that its $(i, j)$ element is $k_f(\mathbf{t}_{nr}[i], \mathbf{t}_{nr'}[j]),$ and similarly define the matrices $\mathbf{K}_{g}(\mathbf{t}_{nr}, \mathbf{t}_{nr'})$ and $\mathbf{K}_{h}(\mathbf{t}_{nr}, \mathbf{t}_{nr'})$. We model the response vector as 

$$ \mathbf{y}_{nr} = \mathbf{f}_{nr} + \mathbf{e},\;\; \mathbf{e} \sim \mathcal{N}(\mathbf{0}, \sigma^2 \mathcal{I}). $$ 

We can see the following conditional distributions,

$$  
\begin{aligned} 
(\mathbf{y}_{nr} | \mathbf{f}_{nr}, \sigma^2) &\sim \mathcal{N}(\mathbf{f}_{nr}, \sigma^2 \mathcal{I}) \\
(\mathbf{f}_{nr} | \mathbf{g}_{n}, \theta_f ) &\sim \mathcal{N}(\mathbf{g}_{n}, \mathbf{K}_{f}(\mathbf{t}_{nr}, \mathbf{t}_{nr})) \\ 
(\mathbf{g}_{n} | \mathbf{h}_{i}, \theta_g ) &\sim \mathcal{N}(\mathbf{h}_{i}, \mathbf{K}_{g}(\mathbf{t}_{nr}, \mathbf{t}_{nr})) \\
(\mathbf{h}_i | \mathbf{t}_{nr}, \theta_h) &\sim \mathcal{N}(\mathbf{0}, \mathbf{K}_h(\mathbf{t}_{nr}, \mathbf{t}_{nr}))
\end{aligned} 
$$

We can see that there are two levels of hierarchy here:

- Replicates within genes
- Genes within clusters

It is easy to find the marginal likelihood of $\bf{y}_{nr}$ through the properties of the multivariate normal distribution,



$$
(\mathbf{y}_{nr} | \mathbf{t}_{nr}, \theta_f, \theta_g, \theta_f) \sim \mathcal{N}(\mathbf{0}, \mathbf{K}_{f}(\mathbf{t}_{nr}, \mathbf{t}_{nr}) + \mathbf{K}_{g}(\mathbf{t}_{nr}, \mathbf{t}_{nr}) + \mathbf{K}_{h}(\mathbf{t}_{nr}, \mathbf{t}_{nr}) + \sigma^2 \mathcal{I}).
$$

## Implementation in R

First we will library in the required packages and read in the data in the format we want, which reflects the hierarchical nature of the data. 

### Read in the data

```{r data, message=F, warning=F}
library(dplyr)
library(forcats)
library(ggplot2)
theme_set(theme_bw(base_size = 12))

drosophila <- read.csv("data/drosophila.csv")

drosophila <- drosophila %>%
  mutate(gene = fct_inorder(gene))
```

### Plot the data

```{r, make_plot}
allgenes <- drosophila %>% 
    ggplot(aes(time, log2exp)) +
    geom_point(aes(col = group, pch = replicate)) +
    facet_wrap(~gene) +
    labs(title = "Expression Profiles for all Genes",
         x = "Time",
         y = "log2-Normalized gene expression") + 
    scale_colour_brewer(palette = "Dark2") +
    scale_x_continuous(breaks = seq(0, 12, by = 2)) 
```


See Figure \@ref(fig:facet1) for example:

```{r facet1, fig.cap='faceted plot of all genes', tidy=FALSE, echo = F}
allgenes
```

## Appendix: R session information

```{r}
sessionInfo()
```

